File: services/tools/workspace.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from services.config import config

MAX_READ_BYTES = 200_000


def workspace_root() -> Path:
    return config.workspace_root


def resolve_path(path: str, *, must_exist: bool = False) -> Path:
    """把相对路径解析到工作区内，禁止越界访问。"""
    root = workspace_root()
    raw = (path or ".").strip()
    candidate = Path(raw).expanduser()
    if not candidate.is_absolute():
        candidate = root / candidate
    resolved = candidate.resolve()
    root_resolved = root.resolve()
    try:
        resolved.relative_to(root_resolved)
    except ValueError as exc:
        raise ValueError(f"路径超出工作区 {root_resolved}: {path}") from exc
    if must_exist and not resolved.exists():
        raise ValueError(f"路径不存在: {resolved}")
    return resolved


def tool_error(message: str, **payload: Any) -> str:
    import json

    return json.dumps({"ok": False, "error": message, **payload}, ensure_ascii=False)


def tool_ok(**payload) -> str:
    import json

    return json.dumps({"ok": True, **payload}, ensure_ascii=False)


def read_text_limited(path: Path, *, max_bytes: int = MAX_READ_BYTES) -> str:
    size = path.stat().st_size
    if size > max_bytes:
        raise ValueError(f"文件过大 ({size} bytes)，上限 {max_bytes}")
    return path.read_text(encoding="utf-8", errors="replace")

File: apps/server/src/tools/workspace.ts
import path from "node:path";
import fs from "node:fs";

import { config } from "../config.js";

const MAX_READ_BYTES = 200_000;

export function workspaceRoot(): string {
  return config.workspaceRoot;
}

export function resolvePath(inputPath: string, mustExist = false): string {
  const root = workspaceRoot();
  const raw = (inputPath || ".").trim();
  const candidate = path.isAbsolute(raw) ? raw : path.join(root, raw);
  const resolved = path.resolve(candidate);
  const rootResolved = path.resolve(root);
  const relative = path.relative(rootResolved, resolved);
  if (relative.startsWith("..") || path.isAbsolute(relative)) {
    throw new Error(`路径超出工作区 ${rootResolved}: ${inputPath}`);
  }
  if (mustExist && !fs.existsSync(resolved)) {
    throw new Error(`路径不存在: ${resolved}`);
  }
  return resolved;
}

export function relPath(absPath: string): string {
  return path.relative(workspaceRoot(), absPath);
}

export function toolError(message: string, extra: Record<string, unknown> = {}): string {
  return JSON.stringify({ ok: false, error: message, ...extra });
}

export function toolOk(payload: Record<string, unknown>): string {
  return JSON.stringify({ ok: true, ...payload });
}

export async function readTextLimited(
  filePath: string,
  maxBytes = MAX_READ_BYTES,
): Promise<string> {
  const { stat, readFile } = await import("node:fs/promises");
  const size = (await stat(filePath)).size;
  if (size > maxBytes) {
    throw new Error(`文件过大 (${size} bytes)，上限 ${maxBytes}`);
  }
  return readFile(filePath, "utf8");
}
